File: Capacitor-CORS2/ios/Plugin/SSEConnection.swift
import Foundation

class SSEConnection: NSObject {
    private let url: String
    private let headers: [String: String]
    private let withCredentials: Bool
    private let reconnectTimeout: Double
    private weak var plugin: CorsBypassPlugin?
    private let connectionId: String
    
    private var urlSessionTask: URLSessionDataTask?
    private var urlSession: URLSession?
    private var reconnectTimer: Timer?
    private var isConnected = false
    
    init(url: String, headers: [String: String], withCredentials: Bool, reconnectTimeout: Double, plugin: CorsBypassPlugin, connectionId: String) {
        self.url = url
        self.headers = headers
        self.withCredentials = withCredentials
        self.reconnectTimeout = reconnectTimeout
        self.plugin = plugin
        self.connectionId = connectionId
        super.init()
    }
    
    func connect() {
        guard let requestUrl = URL(string: url) else {
            plugin?.notifySSEError(connectionId: connectionId, error: "Invalid URL")
            return
        }
        
        var request = URLRequest(url: requestUrl)
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
        
        // Set custom headers
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }
        
        // Configure session
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 0
        config.timeoutIntervalForResource = 0
        
        if withCredentials {
            config.httpCookieAcceptPolicy = .always
            config.httpShouldSetCookies = true
        }
        
        urlSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
        urlSessionTask = urlSession?.dataTask(with: request)
        urlSessionTask?.resume()
    }
    
    func disconnect() {
        reconnectTimer?.invalidate()
        reconnectTimer = nil
        
        urlSessionTask?.cancel()
        urlSessionTask = nil
        
        urlSession?.invalidateAndCancel()
        urlSession = nil
        
        isConnected = false
        plugin?.notifySSEClose(connectionId: connectionId)
    }
    
    private func scheduleReconnect() {
        guard reconnectTimeout > 0 else { return }
        
        reconnectTimer?.invalidate()
        reconnectTimer = Timer.scheduledTimer(withTimeInterval: reconnectTimeout, repeats: false) { [weak self] _ in
            self?.connect()
        }
    }
    
    private func parseSSEData(_ data: String) {
        let lines = data.components(separatedBy: .newlines)
        var eventType: String?
        var eventData: String?
        var eventId: String?
        
        for line in lines {
            if line.isEmpty {
                // Empty line indicates end of event
                if let data = eventData {
                    plugin?.notifySSEMessage(connectionId: connectionId, data: data, id: eventId, type: eventType)
                }
                
                // Reset for next event
                eventType = nil
                eventData = nil
                eventId = nil
            } else if line.hasPrefix("data:") {
                let dataValue = String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces)
                if eventData == nil {
                    eventData = dataValue
                } else {
                    eventData! += "\n" + dataValue
                }
            } else if line.hasPrefix("event:") {
                eventType = String(line.dropFirst(6)).trimmingCharacters(in: .whitespaces)
            } else if line.hasPrefix("id:") {
                eventId = String(line.dropFirst(3)).trimmingCharacters(in: .whitespaces)
            } else if line.hasPrefix("retry:") {
                // Handle retry directive if needed
            }
        }
        
        // Handle case where data doesn't end with empty line
        if let data = eventData {
            plugin?.notifySSEMessage(connectionId: connectionId, data: data, id: eventId, type: eventType)
        }
    }
}

extension SSEConnection: URLSessionDataDelegate {
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void) {
        
        guard let httpResponse = response as? HTTPURLResponse else {
            plugin?.notifySSEError(connectionId: connectionId, error: "Invalid response")
            completionHandler(.cancel)
            return
        }
        
        if httpResponse.statusCode == 200 {
            isConnected = true
            plugin?.notifySSEOpen(connectionId: connectionId)
            completionHandler(.allow)
        } else {
            plugin?.notifySSEError(connectionId: connectionId, error: "HTTP \(httpResponse.statusCode)")
            completionHandler(.cancel)
        }
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        guard let string = String(data: data, encoding: .utf8) else {
            return
        }
        
        parseSSEData(string)
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didCompleteWithError error: Error?) {
        isConnected = false
        
        if let error = error {
            plugin?.notifySSEError(connectionId: connectionId, error: error.localizedDescription)
        }
        
        // Schedule reconnection if not manually disconnected
        if reconnectTimeout > 0 && urlSessionTask != nil {
            scheduleReconnect()
        }
    }
}

File: Capacitor-CORS2/android/src/main/java/com/capacitor/cors/SSEConnection.java
package com.capacitor.cors;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import com.getcapacitor.JSObject;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Iterator;
import java.util.concurrent.TimeUnit;
import okhttp3.*;

public class SSEConnection {
    private static final String TAG = "SSEConnection";
    
    private final String url;
    private final JSObject headers;
    private final boolean withCredentials;
    private final double reconnectTimeout;
    private final CorsBypassPlugin plugin;
    private final String connectionId;
    private final OkHttpClient httpClient;
    
    private Call currentCall;
    private Handler mainHandler;
    private Handler reconnectHandler;
    private boolean isConnected = false;
    private boolean shouldReconnect = true;

    public SSEConnection(String url, JSObject headers, boolean withCredentials, 
                        double reconnectTimeout, CorsBypassPlugin plugin, 
                        String connectionId, OkHttpClient httpClient) {
        this.url = url;
        this.headers = headers;
        this.withCredentials = withCredentials;
        this.reconnectTimeout = reconnectTimeout;
        this.plugin = plugin;
        this.connectionId = connectionId;
        this.httpClient = httpClient;
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.reconnectHandler = new Handler(Looper.getMainLooper());
    }

    public void connect() {
        try {
            Request.Builder requestBuilder = new Request.Builder()
                .url(url)
                .addHeader("Accept", "text/event-stream")
                .addHeader("Cache-Control", "no-cache");

            // Add custom headers
            Iterator<String> headerKeys = headers.keys();
            while (headerKeys.hasNext()) {
                String key = headerKeys.next();
                requestBuilder.addHeader(key, headers.getString(key));
            }

            Request request = requestBuilder.build();

            // Configure client for SSE
            OkHttpClient client = httpClient.newBuilder()
                .connectTimeout(0, TimeUnit.SECONDS)
                .readTimeout(0, TimeUnit.SECONDS)
                .writeTimeout(0, TimeUnit.SECONDS)
                .build();

            currentCall = client.newCall(request);
            currentCall.enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    if (!call.isCanceled()) {
                        mainHandler.post(() -> {
                            plugin.notifySSEError(connectionId, e.getMessage());
                            scheduleReconnect();
                        });
                    }
                }

                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    if (response.code() == 200) {
                        isConnected = true;
                        mainHandler.post(() -> plugin.notifySSEOpen(connectionId));
                        
                        // Start reading the stream
                        readEventStream(response);
                    } else {
                        mainHandler.post(() -> {
                            plugin.notifySSEError(connectionId, "HTTP " + response.code());
                            scheduleReconnect();
                        });
                    }
                }
            });

        } catch (Exception e) {
            plugin.notifySSEError(connectionId, "Failed to connect: " + e.getMessage());
        }
    }

    private void readEventStream(Response response) {
        try (InputStream inputStream = response.body().byteStream();
             BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
            
            String line;
            StringBuilder eventData = new StringBuilder();
            String eventType = null;
            String eventId = null;
            
            while ((line = reader.readLine()) != null && !Thread.currentThread().isInterrupted()) {
                if (line.isEmpty()) {
                    // Empty line indicates end of event
                    if (eventData.length() > 0) {
                        final String data = eventData.toString();
                        final String type = eventType;
                        final String id = eventId;
                        
                        mainHandler.post(() -> 
                            plugin.notifySSEMessage(connectionId, data, id, type)
                        );
                    }
                    
                    // Reset for next event
                    eventData.setLength(0);
                    eventType = null;
                    eventId = null;
                    
                } else if (line.startsWith("data:")) {
                    String dataValue = line.substring(5).trim();
                    if (eventData.length() > 0) {
                        eventData.append("\n");
                    }
                    eventData.append(dataValue);
                    
                } else if (line.startsWith("event:")) {
                    eventType = line.substring(6).trim();
                    
                } else if (line.startsWith("id:")) {
                    eventId = line.substring(3).trim();
                    
                } else if (line.startsWith("retry:")) {
                    // Handle retry directive if needed
                }
            }
            
        } catch (IOException e) {
            if (!Thread.currentThread().isInterrupted()) {
                mainHandler.post(() -> {
                    plugin.notifySSEError(connectionId, "Stream reading error: " + e.getMessage());
                    scheduleReconnect();
                });
            }
        } finally {
            isConnected = false;
        }
    }

    public void disconnect() {
        shouldReconnect = false;
        reconnectHandler.removeCallbacksAndMessages(null);
        
        if (currentCall != null) {
            currentCall.cancel();
            currentCall = null;
        }
        
        isConnected = false;
        mainHandler.post(() -> plugin.notifySSEClose(connectionId));
    }

    private void scheduleReconnect() {
        if (shouldReconnect && reconnectTimeout > 0) {
            reconnectHandler.postDelayed(() -> {
                if (shouldReconnect) {
                    connect();
                }
            }, (long) (reconnectTimeout * 1000));
        }
    }
}
